### PR TITLE
fix(typescript-estree): account for namespace nesting in AST conversion

### LINKED ISSUE #4966: Bug: incorrect AST produced for nested namespaces
This was forgotten when we worked around #2573

For `namespace foo { namespace bar {} }` we produce the following AST:
`TSModuleDeclaration[id = foo] > TSModuleBlock > TSModuleDeclaration[id = bar] > TSModuleBlock`

For `namespace foo.bar {}` produces the following AST:
`TSModuleDeclaration[id = foo] > TSModuleDeclaration[id = bar] > TSModuleBlock`

The AST we produce is driven by the AST TS produces - they emit the nested module declaration, so our code does as well.
Ultimately however is pretty incorrect and does not correctly represent the code as written.

Instead we should emit an `EntityName`, thus allowing `TSQualifiedName` in the `.id`:
`TSModuleDeclaration[id = foo.bar] > TSModuleBlock`

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #4966
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

TypeScript produces a nested body AST for Identifier names like `abc.def`. This unravels their body blocks and accumulates the name if necessary.

### CODE REVIEW

File: packages/typescript-estree/src/convert.ts [hunk 1]
```diff
@@ -735,6 +731,41 @@ export class Converter {
...
+        node.body as ts.ModuleBlock | undefined,
+        this.convertChild(node.name),
+      ];
+    }
+
+    let name = node.name.text;
+
+    while (node.body && ts.isModuleDeclaration(node.body)) {
+      node = node.body;
+      name = `${name}.${node.name.text}`;
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/convert.ts:756]: #6116 🤷

File: packages/typescript-estree/src/convert.ts [hunk 2]
```diff
@@ -254,12 +254,8 @@ export class Converter {
...
   ): T {
     const result = data;
-    if (!result.range) {
-      result.range = getRange(node, this.ast);
-    }
-    if (!result.loc) {
-      result.loc = getLocFor(result.range[0], result.range[1], this.ast);
-    }
+    result.range ??= getRange(node, this.ast);
+    result.loc ??= getLocFor(result.range[0], result.range[1], this.ast);
```
Comment (JoshuaKGoldberg) [inline packages/typescript-estree/src/convert.ts:258]: Couldn't resist while I was here 😄 

Feels like this should be a lint rule enabled always...

**Review by bradzacher — CHANGES_REQUESTED** [review]
I do wish we could use `Identifier` for the dotted name - unfortunately `Identifier.name` needs to be a valid JS identifier - dots aren't allowed!

Instead we can convert this to a `TSQualifiedName`

**Comment by bradzacher** [comment]
Now that #6443 has landed in main, we'll need to merge v6 then update this PR accordingly.

**Review by bradzacher — APPROVED** [review]
all_green

**Pull request merged**